File: topic2_lists/InsertNodeAtTail.java
// SpaniardsToSv. Please report an issue if you have any questions.

// HackerRank documentation.
/*
  Insert Node at the end of a linked list
  head pointer input could be NULL as well for empty list
  Node is defined as
  
  class Node {
     int data;
     Node next;
  }
*/

/**
* Our solution. Please note that there could always be more than one.
*
*/
Node Insert(Node head,int data) {
    Node tail = new Node();
    tail.data = data;
    if (head == null) {
        return tail;
    }

    Node trackingNode = head;

    while (trackingNode.next != null) {
        trackingNode = trackingNode.next;
    }
    trackingNode.next = tail;
    return head;
}

File: topic2_lists/js/InsertNodeAtTail.js
// https://www.hackerrank.com/challenges/insert-a-node-at-the-tail-of-a-linked-list/problem

function insertNodeAtTail(head, data) {
  let it = head;
  
  // go till the last node
  while (it && it.next) {
      it = it.next;
  }
  
  let node = new SinglyLinkedListNode(data);
  
  if (it) it.next = node;
  else head = node;
  
  return head;
}
